File: zh-Hant/codes/dart/chapter_sorting/merge_sort.dart
/**
 * File: merge_sort.dart
 * Created Time: 2023-02-14
 * Author: what-is-me (whatisme@outlook.jp)
 */

/* 合併左子陣列和右子陣列 */
void merge(List<int> nums, int left, int mid, int right) {
  // 左子陣列區間為 [left, mid], 右子陣列區間為 [mid+1, right]
  // 建立一個臨時陣列 tmp ，用於存放合併後的結果
  List<int> tmp = List.filled(right - left + 1, 0);
  // 初始化左子陣列和右子陣列的起始索引
  int i = left, j = mid + 1, k = 0;
  // 當左右子陣列都還有元素時，進行比較並將較小的元素複製到臨時陣列中
  while (i <= mid && j <= right) {
    if (nums[i] <= nums[j])
      tmp[k++] = nums[i++];
    else
      tmp[k++] = nums[j++];
  }
  // 將左子陣列和右子陣列的剩餘元素複製到臨時陣列中
  while (i <= mid) {
    tmp[k++] = nums[i++];
  }
  while (j <= right) {
    tmp[k++] = nums[j++];
  }
  // 將臨時陣列 tmp 中的元素複製回原陣列 nums 的對應區間
  for (k = 0; k < tmp.length; k++) {
    nums[left + k] = tmp[k];
  }
}

/* 合併排序 */
void mergeSort(List<int> nums, int left, int right) {
  // 終止條件
  if (left >= right) return; // 當子陣列長度為 1 時終止遞迴
  // 劃分階段
  int mid = (left + right) ~/ 2; // 計算中點
  mergeSort(nums, left, mid); // 遞迴左子陣列
  mergeSort(nums, mid + 1, right); // 遞迴右子陣列
  // 合併階段
  merge(nums, left, mid, right);
}

/* Driver Code */
void main() {
  /* 合併排序 */
  List<int> nums = [7, 3, 2, 6, 0, 1, 5, 4];
  mergeSort(nums, 0, nums.length - 1);
  print("合併排序完成後 nums = $nums");
}

File: ja/codes/ruby/chapter_sorting/merge_sort.rb
=begin
File: merge_sort.rb
Created Time: 2024-04-10
Author: junminhong (junminhong1110@gmail.com)
=end

### 左部分配列と右部分配列をマージ ###
def merge(nums, left, mid, right)
  # 左部分配列の区間は [left, mid]、右部分配列の区間は [mid+1, right]
  # マージ結果を格納する一時配列 tmp を作成
  tmp = Array.new(right - left + 1, 0)
  # 左右の部分配列の開始インデックスを初期化する
  i, j, k = left, mid + 1, 0
  # 左右の部分配列にまだ要素がある間は比較し、小さいほうを一時配列にコピーする
  while i <= mid && j <= right
    if nums[i] <= nums[j]
      tmp[k] = nums[i]
      i += 1
    else
      tmp[k] = nums[j]
      j += 1
    end
    k += 1
  end
  # 左右の部分配列の残り要素を一時配列にコピーする
  while i <= mid
    tmp[k] = nums[i]
    i += 1
    k += 1
  end
  while j <= right
    tmp[k] = nums[j]
    j += 1
    k += 1
  end
  # 一時配列 tmp の要素を元の配列 nums の対応区間にコピーする
  (0...tmp.length).each do |k|
    nums[left + k] = tmp[k]
  end
end

### マージソート ###
def merge_sort(nums, left, right)
  # 終了条件
  # 部分配列の長さが 1 になったら再帰を終了する
  return if left >= right
  # 分割フェーズ
  mid = left + (right - left) / 2 # 中点を計算
  merge_sort(nums, left, mid) # 左部分配列を再帰処理
  merge_sort(nums, mid + 1, right) # 右部分配列を再帰処理
  # マージフェーズ
  merge(nums, left, mid, right)
end

### Driver Code ###
if __FILE__ == $0
  nums = [7, 3, 2, 6, 0, 1, 5, 4]
  merge_sort(nums, 0, nums.length - 1)
  puts "マージソート完了後 nums = #{nums.inspect}"
end
